subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/resolveengine/graph/builder/DependencyGraphBuilder.java

```diff
@@ -136,13 +152,17 @@ private void traverseGraph(final ResolveState resolveState) {
                 resolveEdges(node, dependencies, dependenciesMissingLocalMetadata, resolveState, componentIdentifierCache);
             } else {
                 // We have some batched up conflicts. Resolve the first, and continue traversing the graph
-                conflictHandler.resolveNextConflict(resolveState.getReplaceSelectionWithConflictResultAction());
+                if (moduleConflictHandler.hasConflicts()) {
+                    moduleConflictHandler.resolveNextConflict(resolveState.getReplaceSelectionWithConflictResultAction());
+                } else {
+                    capabilitiesConflictHandler.resolveNextConflict(resolveState.getReplaceSelectionWithConflictResultAction());
+                }
             }
 
         }
     }
 
-    private void performSelection(final ResolveState resolveState, ComponentState moduleRevision) {
+    private void performSelection(final ResolveState resolveState, final ComponentState moduleRevision) {
```

Comment:
Leftover from previous spike?

Reply:
Yes, and no :) The `final` was required by an initial implementation which used an inner class (this wasn't ever committed). I could have removed it. Then I could also have removed the `final` on `ResolveState` which is not needed either.